src/classes/OPP_CampaignMember_TDTM.cls

```diff
@@ -0,0 +1,126 @@
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.com Foundation
+* @date 2015
+* @group Opportunity
+* @group-content ../../ApexDocContent/Opportunities.htm
+* @description Handles adding and updating campaign members based on campaign donations.
+*/
+public with sharing class OPP_CampaignMember_TDTM extends TDTM_Runnable {
+    
+    /**@description Lists to pass to CON_AddToCampaign*/
+    private list<Id> listContactId = new list<Id>();
+    private list<Id> listCampaignId = new list<Id>();
+    private list<String> listMemberStatus = new list<String>();
+
+    /**@description Static flag to prevent recursion. */
+    public static boolean hasRun = false;
+
+    /*******************************************************************************************************
+    * @description Creates and updates campaign members based on opportunity primary contact, primary 
+    * campaign source, and the primary contact status field.
+    * @param listNew the list of Opportunities from trigger new. 
+    * @param listOld the list of Opportunities from trigger old. 
+    * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.). 
+    * @param objResult the describe for Opportunities.
+    * @return dmlWrapper Campaign Members to insert or update.
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+        TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+        
+        dmlWrapper dmlWrapper = new DmlWrapper();
+
+        if (hasRun)
+            return dmlWrapper;
+            
+        for (integer i=0; i<newlist.size(); i++) {
+
+            Opportunity opp = (Opportunity)newlist[i];
+            
+            //if status is "omit", or the primary contact or campaign are blank,
+            //continue to the next opportunity without doing any campaign member work
+            if ((!string.isBlank(opp.Primary_Contact_Campaign_Member_Status__c) && 
+                    opp.Primary_Contact_Campaign_Member_Status__c.equalsIgnoreCase(Label.CampaignMemberStatusOmit)) ||
+                    opp.Primary_Contact__c == null || opp.CampaignId == null) {
+
+                continue;
+            }
+
+            if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
+                addOppMemberFields(opp);
+            }
+            
+            if (triggerAction == TDTM_Runnable.Action.AfterUpdate) {
+                Opportunity oldOpp = (Opportunity)oldlist[i];
+
+                //the status is blank and we're changing to or from closed/won
+                if (string.isBlank(opp.Primary_Contact_Campaign_Member_Status__c) && 
+                    (opp.IsClosed != oldOpp.isClosed && opp.IsWon != oldOpp.isWon)) {
```

Comment:
should this be changing to/from Close OR to/from Won, rather than requiring both?